lib/session.ts

```diff
@@ -327,20 +335,65 @@ export class Session extends EventEmitter {
 
     for (const eventName in SessionEvents) {
       this._session.on(SessionEvents[eventName],
-        (context) => this.emit(SessionEvents[eventName], EventContext.translate(context, this)));
+        (context) => {
+          const params: EmitParameters = {
+            rheaContext: context,
+            emitter: this,
+            eventName: SessionEvents[eventName],
+            emitterType: "session",
+            connectionId: this.connection.id
+          };
+          emitEvent(params);
+        });
     }
 
     // Add event handlers for *_error and *_close events that can be propogated to the session
     // object, if they are not handled at their level. * denotes - Sender and Receiver.
+
     // Sender
-    this._session.on(rhea.SenderEvents.senderError,
-      (context) => this.emit(rhea.SenderEvents.senderError, EventContext.translate(context, this)));
-    this._session.on(rhea.SenderEvents.senderClose,
-      (context) => this.emit(rhea.SenderEvents.senderClose, EventContext.translate(context, this)));
+    this._session.on(SenderEvents.senderError, (context) => {
+      const params: EmitParameters = {
+        rheaContext: context,
+        emitter: this,
+        eventName: SenderEvents.senderError,
+        emitterType: "session",
+        connectionId: this.connection.id
+      };
+      emitEvent(params);
+    });
+    this._session.on(SenderEvents.senderClose, (context) => {
+      const params: EmitParameters = {
+        rheaContext: context,
+        emitter: this,
+        eventName: SenderEvents.senderClose,
+        emitterType: "session",
+        connectionId: this.connection.id
+      };
+      emitEvent(params);
+    });
+
     // Receiver
-    this._session.on(rhea.ReceiverEvents.receiverError,
-      (context) => this.emit(rhea.ReceiverEvents.receiverError, EventContext.translate(context, this)));
-    this._session.on(rhea.ReceiverEvents.receiverClose,
-      (context) => this.emit(rhea.ReceiverEvents.receiverClose, EventContext.translate(context, this)));
+    this._session.on(ReceiverEvents.receiverError, (context) => {
+      const params: EmitParameters = {
+        rheaContext: context,
+        emitter: this,
+        eventName: ReceiverEvents.receiverError,
+        emitterType: "session",
+        connectionId: this.connection.id
+      };
+      emitEvent(params);
+    });
+    this._session.on(ReceiverEvents.receiverClose, (context) => {
+      const params: EmitParameters = {
```

Comment:
why these new params objects everywhere?

Reply:
Rather than having the same emit call and log statement everywhere, this function encapsulates that and also handles setTimeout for errors. Didn't want to repeat there for every event listener being added.